MINOR: trivial cleanups, javadoc errors, omitted StateStore tests, etc.

dongjinleekr opened the pull request:
- Add omitted `[WindowStoreBuilderTest, TimestampedWindowStoreBuilderTest]#shouldThrowNullPointerIfMetricsScopeIsNull`: Other `StateStore`s has it.
- Improve `Stores` Javadoc
- Remove unused method + duplicated parameters
- Remove duplicated spaces
- Remove unthrown `Exception`s + align javadoc parameters
- Remove unnecessary `public` from `[Sink,Source]TaskContext`
- Add omitted `WindowStore`, `SessionStore` test cases: `GlobalStateStoreProviderTest`, `StreamThreadStateStoreProviderTest` (Compare it with `KeyValueStore` counterpart cases.)
- Fix `MeteredTimestampedWindowStore` javadoc

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dongjinleekr commented:
Retest this please.

dongjinleekr commented:
Retest this please.

dongjinleekr commented:
Rebased onto the latest trunk & remove one more unused import. cc/ @hachikuji @omkreddy

dongjinleekr commented:
Retest this please.

dongjinleekr commented:
Rebased onto the latest trunk, with removing duplicated entity in `checkstyle/import-control.xml`.

omkreddy commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -807,7 +807,7 @@ public void abortTransaction() throws ProducerFencedException {
      * </p>
      * <p>
-     * Some transactional send errors cannot be resolved with a call to {@link #abortTransaction()}.  In particular,
+     * Some transactional send errors cannot be resolved with a call to {@link #abortTransaction()}. n particular,
```
n -> "In"

dongjinleekr replied:
Fixed! :smiley:

omkreddy commented:
retest this please

omkreddy commented:
cc @mjsax @vvcephei for streams test changes

mjsax commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -47,7 +47,7 @@
  * response.
  *
- * FetchSessionHandler tracks the partitions which are in the session.  It also
+ * FetchSessionHandler tracks the partitions which are in the session. It also
```
Just to clarify. To put two blanks between two sentences is a writing style that goes back to classic type writers -- having a larger space between sentences makes it easier to read. Hence, removing all those double-blanks (to correct writing style) is just a lot of "noise" with no real value...

We can still merge this PR (as you put a lot of work into it already), however, I would encourage you to not do such PRs in the future. It's just additional review load (and committers already cannot handle the current load...) with no real value and thus a "waste" of your own time as well as the reviewer's time.

dongjinleekr replied:
@mjsax I see. I’ll be more careful from now on. Thanks for the detailed explanation.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/WindowedSerdes.java:
```diff
@@ -64,14 +64,6 @@ public SessionWindowedSerde(final Serde<T> inner) {
-     * Construct a {@code TimeWindowedSerde} object to deserialize changelog topic
-     * for the specified inner class type and window size.
-     */
-    static public <T> Serde<Windowed<T>> timeWindowedSerdeFrom(final Class<T> type, final long windowSize) {
```
This is public API and cannot just be removed. Why do you think it should be removed?

dongjinleekr replied:
I'm sorry, it was a mistake; So reverted.

mjsax replied:
@dongjinleekr Seems you did not revert this change yet.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/TimestampedWindowStoreBuilderTest.java:
```diff
@@ -179,4 +179,9 @@ public void shouldThrowNullPointerIfTimeIsNull() {
     }
 
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowNullPointerIfMetricsScopeIsNull() {
```
Where do we pass a `null` value? Also, where is the `MetricScope` in this test?

dongjinleekr replied:
`supplier#metricsScope()` returns `null`, for its `setUp()` configuration. It is same to `[KeyValueStoreBuilderTest, TimestampedKeyValueStoreBuilderTest]#shouldThrowNullPointerIfMetricsScopeIsNull`. `SessionStoreBuilderTest#shouldThrowNullPointerIfMetricsScopeIsNull` does same, but it also validates Exception message. (Wait, is this way more recommended?)

mjsax replied:
Thanks for the clarification. However, I am still not sure if I understand this test. The constructor `TimestampedWindowStoreBuilder` does not call `supplier.metricsScope()` -- hence, it am still puzzled why this test would throw a NPE? To test `metricScope()` we would need to call `build()` method. In fact, following the call trace, we would never check if `metricScope()` returns `null` or not. Seems worth to fix... (maybe add a check in `MeteredWindowStore` constructor? -- not sure if the other `MeteredXxxStore` have a similar check. If not, would be good to add one.)

Seems it comes from `supplier.name()` -- for this case, the test setup is incorrect -- also in the test above actually, that would also fail with `supplier.name()` NPE, but not with the NPE they want to test.

> does same, but it also validates Exception message. (Wait, is this way more recommended?)

Yes, it's always recommended to verify the exception message to make sure we test the right think. Using `(expected = NullPointerException.class)` annotation is not ideal.

dongjinleekr replied:
@mjsax I reviewed the problem more carefully and concluded that you are right - the `NullPointerException` here is not related to `MetricScope`. We need to rewrite the series of tests.

Let's review `TimestampedWindowStoreBuilderTest#shouldThrowNullPointerIfMetricsScopeIsNull` as an example. Here, `supplier#name` is `null` so when instantiating `TimestampedWindowStoreBuilder`, it fails in its parent class's constructor, `AbstractStoreBuilder` line 41 - `Objects.requireNonNull(name, "name cannot be null");`. It is also why `SessionStoreBuilderTest#shouldThrowNullPointerIfMetricsScopeIsNull` ends with a error message "name cannot be null" - obviously, it also has nothing to do with `MetricScope`.

Here is the debugger status; As you can see, `supplier#name` is `null` here.

![name-null](https://user-images.githubusercontent.com/2375128/79251817-a75b8b00-7ebb-11ea-9b1e-550d203c82df.png)

I am now rewriting the tests. Stay tuned! :smile:

mjsax replied:
Thanks! Highly appreciated!

mjsax commented on streams/src/test/java/org/apache/kafka/streams/state/internals/WindowStoreBuilderTest.java:
```diff
@@ -134,4 +134,9 @@ public void shouldThrowNullPointerIfTimeIsNull() {
     }
 
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowNullPointerIfMetricsScopeIsNull() {
```
Same question as above?

dongjinleekr replied:
Same answer as above. :sweat:

mjsax commented:
The build is currently broker. There is already a PR to fix it: https://github.com/apache/kafka/pull/8424

No need to retest until `trunk` is fixed.

mjsax commented:
@dongjinleekr -- this PR needs a rebase. Might be nice to get it into 2.6?

dongjinleekr committed:
Add omitted [WindowStoreBuilderTest, TimestampedWindowStoreBuilderTest]#shouldThrowNullPointerIfMetricsScopeIsNull: Other StateStores has it.

dongjinleekr committed:
Improve Stores Javadoc

1. Fix javadoc typo: Stores#timestampedWindowStoreBuilder, windows-store -> window-store (Compare with Stores#timestampedKeyValueStoreBuilder)
2. Add omitted IllegalArgumentException condition in Stores#lruMap.
3. Add omitted IllegalArgumentException condition in WindowBytesStoreSupplier methods.

dongjinleekr committed:
Remove unused method + duplicated parameters

1. Remove unused method: TimeWindowedSerde#timeWindowedSerdeFrom(Class, long)
2. Remove duplicated parameter: Log#splitOverflowedSegment

dongjinleekr committed:
Remove unthrown Exceptions: TaskManager, ProduceBenchWorker

dongjinleekr committed:
Remove unnecessary public from [Sink,Source]TaskContext

dongjinleekr committed:
Fix incorrect MeteredTimestampedWindowStore javadoc

dongjinleekr committed:
Remove duplicated entity in checkstyle/import-control.xml

dongjinleekr committed:
Add omitted WindowStore, SessionStore test cases: GlobalStateStoreProviderTest, StreamThreadStateStoreProviderTest (Note: compare with KeyValueStore cases.)

dongjinleekr committed:
Update [KeyValue,TimestampedKeyValue,WindowStore,TimestampedWindowStore]BuilderTest#shouldThrowNullPointerIfMetricsScopeIsNull: now checks Exception message.

dongjinleekr committed:
Give BytesStoreSupplier mock with name to [KeyValue,TimestampedKeyValue,WindowStore,TimestampedWindowStore]BuilderTest#shouldThrowNullPointerIfMetricsScopeIsNull; No exceptions are thrown now.

dongjinleekr committed:
Add metricsScope nullity check to [KeyValue,TimestampedKeyValue,Window,TimestampedWindow,Session]StoreBuilder constructors

dongjinleekr committed:
Add nullity check in 'WindowStoreBuilder' constructor. (For parity with 'TimestampedWindowStoreBuilder', 'SessionStoreBuilder', etc.)

dongjinleekr committed:
Remove unthrown Exception in SessionStoreBuilderTest#setUp

dongjinleekr committed:
Fix typo in `XXXStoreBuilder` validation methods

- `KeyValueStoreBuilder`: `bytesStoreSupplier` → `storeSupplier`
- `TimestampedKeyValueStoreBuilder`: `bytesStoreSupplier` → `storeSupplier`
- `TimestampedWindowStoreBuilder`: `bytesStoreSupplier` → `storeSupplier`
- `SessionStoreBuilder`: `supplier` → `storeSupplier`

dongjinleekr committed:
Remove unused variables: RocksDBStore, SuppressionIntegrationTest

dongjinleekr committed:
Remove redundant type parameters from AbstractJoinIntegrationTest, SuppressScenarioTest

dongjinleekr commented:
@mjsax Here is the fix. I investigated the problem and found the following.

The root of the problem is: the consturctors of `XXXStoreBuilder` classes omit the MetricsScope nullity check (see commit `5b432d9`.) As of present, the `NullPointerException`s are thrown from the constructor of `AbstractStoreBuilder` - it has nothing to do with MetricsScope. (Interestingly, `SessionStoreBuilderTest#shouldThrowNullPointerIfMetricsScopeIsNull` already takes the right way.)

As a evidence, if we give the `XXXBytesStoreSupplier` mock with non-null `name` no exception is thrown. (see commit `01f2cf3`)

To fix this problem, I added a validation method to each of the constructor of `XXXStoreBuilder`; To make the tests not broken, I also made the supplier mocks to return non-null MetricsScope by `expect(supplier.metricsScope()).andReturn("metricScope");` (see commit `2e0043b`)

+1. I removed the 'two space' removal while rebasing onto the latest trunk; You are right, it is worthless. Instead, I added the other trivial fixes I found during reading the code.

dongjinleekr commented:
@mjsax Could this PR get it into 2.7? :smiley:

mjsax commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -2419,7 +2419,7 @@ class Log(@volatile private var _dir: File,
       // replace old segment with new ones
       info(s"Replacing overflowed segment $segment with split segments $newSegments")
-      replaceSegments(newSegments.toList, List(segment), isRecoveredSwapFile = false)
+      replaceSegments(newSegments.toList, List(segment))
```
@guozhangwang Is this ok? I am not familiar with this part of the code. If `false` the default anyway?

guozhangwang replied:
LGTM, since the method definition also has `isRecoveredSwapFile: Boolean = false` by default.

mjsax replied:
Thanks @guozhangwang for verifying.

mjsax reviewed (commented):
Overall LGTM.

I am not familiar with `core` code so I hope @guozhangwang helps out. One comment though that needs to be resolved.

mjsax commented:
It also seems that some test failed due to NPE. Those seems to need updates.

dongjinleekr committed:
Revert WindowedSerdes#timeWindowedSerdeFrom(Class, long)

dongjinleekr committed:
Add supplier mock object in TimestampedKeyValueStoreMaterializerTest#shouldCreateKeyValueStoreWithTheProvidedInnerStore to return metricScope: without it, NullPointerExcepion is thrown during TimestampedKeyValueStore instance creation.

dongjinleekr commented:
@mjsax The failing test is now fixed with commit 9f7a73f and mistakenly removed method is reverted in commit 6972309. Also, rebased onto the latest trunk.
@guozhangwang Thanks for reviewing the code! :smiley:

mjsax commented:
Thanks for the PR @dongjinleekr! Merged to `trunk`.
